service-registry/src/main/java/io/servicecomb/serviceregistry/consumer/MicroserviceVersions.java

```diff
@@ -0,0 +1,178 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.servicecomb.serviceregistry.consumer;
+
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.eventbus.Subscribe;
+
+import io.servicecomb.serviceregistry.RegistryUtils;
+import io.servicecomb.serviceregistry.api.registry.MicroserviceInstance;
+import io.servicecomb.serviceregistry.api.response.MicroserviceInstanceChangedEvent;
+import io.servicecomb.serviceregistry.definition.DefinitionConst;
+import io.servicecomb.serviceregistry.task.event.PullMicroserviceVersionsInstancesEvent;
+
+public class MicroserviceVersions {
+  private static final Logger LOGGER = LoggerFactory.getLogger(MicroserviceVersions.class);
+
+  private AppManager appManager;
+
+  private String appId;
+
+  private String microserviceName;
+
+  private List<MicroserviceInstance> instances;
+
+  // key is service id
+  private Map<String, MicroserviceVersion> versions = new ConcurrentHashMap<>();
+
+  // key is version rule
+  private Map<String, MicroserviceVersionRule> versionRules = new ConcurrentHashMap<>();
+
+  // process pulled instances and create versionRule must be protected by lock
+  // otherwise maybe lost instance or version in verisonRule
+  private final Object lock = new Object();
+
+  // to avoid pull too many time
+  // only pendingPullCount is 0, then do a real pull 
+  private AtomicInteger pendingPullCount = new AtomicInteger();
+
+  public MicroserviceVersions(AppManager appManager, String appId, String microserviceName) {
+    this.appManager = appManager;
+    this.appId = appId;
+    this.microserviceName = microserviceName;
+
+    LOGGER.info("create MicroserviceVersions, appId={}, microserviceName={}.",
+        appId,
+        microserviceName);
+
+    appManager.getEventBus().register(this);
+  }
+
+  public String getAppId() {
+    return appId;
+  }
+
+  public String getMicroserviceName() {
+    return microserviceName;
+  }
+
+  public Map<String, MicroserviceVersion> getVersions() {
+    return versions;
+  }
+
+  public void submitPull() {
+    pendingPullCount.incrementAndGet();
+    pullInstances();
+  }
+
+  public void pullInstances() {
+    if (pendingPullCount.decrementAndGet() != 0) {
+      return;
+    }
+
+    List<MicroserviceInstance> pulledInstances = RegistryUtils.findServiceInstance(appId,
+        microserviceName,
+        DefinitionConst.VERSION_RULE_ALL);
+    if (pulledInstances == null) {
+      return;
+    }
+
+    setInstances(pulledInstances);
+  }
+
+  private void setInstances(List<MicroserviceInstance> pulledInstances) {
+    synchronized (lock) {
```

Comment:
Do we have cach or copy on write list for it?
I'm not a big fan of synchronized.

Reply:
can not use that 
as i comment in field , what we protected is not one container, but  is "process pulled instances and create versionRule"